test/unit/common/test_shardtrie.py

```diff
@@ -0,0 +1,312 @@
+                            "parent": "b",
+                            "timestamp": "0000000002.00000",
+                            "flag": 1,
+                            "key": "d",
+                            "data": {
+                                "etag": "x",
+                                "content_type": "text/plain",
+                                "size": 0
+                            },
+                            "children": []
+                        }
+                    ]
+                }
+            ],
+            "metadata": {
+                "data_node_count": 3
+            }
+        })
+
+    def setUp(self):
+        self.trie = shardtrie.ShardTrie.load_from_json(self._test_shard_trie)
+
+    def tearDown(self):
+        pass
+
+    def test_shardtrie_split(self):
+        """Test split trie by full key"""
+        trie = self.trie
+
+        subtrie = trie.split_trie('ab')
+        self.assertTrue(trie.get_node('ab').is_distributed)
+        self.assertEqual(subtrie.root.parent, None)
+        self.assertEqual(len(list(trie.get_data_nodes())), 1)
+        self.assertEqual(len(list(subtrie.get_data_nodes())), 2)
+
+    def test_shardtrie_split_nonexistant_key(self):
+        """Test split trie with non-existant key"""
+        trie = self.trie
+
+        with self.assertRaises(shardtrie.ShardTrieException):
+            subtrie = trie.split_trie('no-such-key')
+
+    def test_shardtrie_get_node(self):
+        """Test getting a node"""
+        trie = self.trie
+
+        node_abc = trie.get_node('abc')
+        self.assertEqual(node_abc.flag, shardtrie.DATA_PRESENT)
+
+    def test_shardtrie_get_distributed_node(self):
+        """Test getting a node from a distributed branch"""
+        trie = self.trie
+        subtrie = trie.split_trie('ab')
+
+        with self.assertRaises(shardtrie.ShardTrieDistributedBranchException):
+            trie.get_node('abc')
+
+    def test_shardtrie_join(self):
+        """Test joining trie"""
+        trie = self.trie
+
+        subtrie = trie.split_trie('ab')
+        self.assertEqual(len(list(trie.get_data_nodes())), 1)
+
+        trie.join_subtrie(subtrie)
+        self.assertEqual(len(list(trie.get_data_nodes())), 3)
+
+    def test_shardtrie_join_subtrie_not_trie(self):
+        """Test joining trie when subtrie is not an instance of ShardTrie"""
+        trie = self.trie
+        subtrie = trie.split_trie('ab')
+
+        with self.assertRaises(shardtrie.ShardTrieException):
+            trie.join_subtrie({'this_is': 'invalid'})
+
+    def test_shardtrie_popping_iterator(self):
+        """Test iterating across the trie while removing visited elements"""
+        trie = self.trie
+
+        list(trie.root.popping_iter())
+        # We cannot use the following line as is_empty checks that the key
+        #  is empty string, popping_iterator does not set the key to ''
+        # self.assertTrue(trie.is_empty())
+        self.assertEqual(trie.root.children, {})
```

Comment:
Popping iterater was an attempt at being more memory efficient when getting.. it isn't used.. but might still be the right idea. This is where benchmarking will come in to play.